### ISSUE TITLE
Allow choice of autowiring empty collections instead of null (in prototype beans) [SPR-10479]

### ISSUE THREAD

**Issue by spring-issuemaster** [spring-projects/spring-framework#15112]
**[John Hendrikx](https://jira.spring.io/secure/ViewProfile.jspa?name=john16384)** opened **[SPR-10479](https://jira.spring.io/browse/SPR-10479?redirect=false)** and commented

I've been looking for hours on how to possibly get Spring to wire up a collection as an empty collection instead of null when no matching Beans are found for the collection type.  The documentation hints at being able to replace the default CustomCollectionEditors (which have a tantalizing "nullAsEmptyCollection" constructor parameter), but so far, none of my replacement attempts have resulted in my PropertyEditor actually being used during autowiring (it's instantiated, but not used).

Things I tried:

```
((ConfigurableBeanFactory)applicationContext.getBeanFactory()).registerCustomEditor(Set.class, DefaultCustomCollectionEditor.class);
```

...ends up being instantiated (several times) but not being used for the actual wiring.

Another thing I tried:

```
  @Bean(name = "customEditorConfigurer")
  public CustomEditorConfigurer getCustomEditorConfigurer() {
    CustomEditorConfigurer configurer = new CustomEditorConfigurer();

    configurer.setCustomEditors(new HashMap<String, Object>() {{
      put("java.util.Set", "hs.dm.DefaultCustomCollectionEditor");
    }});

    configurer.setOrder(Integer.MIN_VALUE);

    return configurer;
  }
```

... again, instantiated, but not used.

I've tried seeing if there is anyway to get a hold of the BeanWrapperImpl class, which has a "overrideDefaultEditor" method in its PropertyEditorRegistrySupport parent class, but I don't know how to reach it.  The source code for PropertyEditorRegistrySupport in #createDefaultEditors again hints at the option to replace them, but I cannot find how:

```
// Default instances of collection editors.
// Can be overridden by registering custom instances of those as custom editors.
this.defaultEditors.put(Collection.class, new CustomCollectionEditor(Collection.class));
this.defaultEditors.put(Set.class, new CustomCollectionEditor(Set.class));
this.defaultEditors.put(SortedSet.class, new CustomCollectionEditor(SortedSet.class));
this.defaultEditors.put(List.class, new CustomCollectionEditor(List.class));
this.defaultEditors.put(SortedMap.class, new CustomMapEditor(SortedMap.class));
```

Section 7.4.2 "Built-in PropertyEditor implementations" of the manual makes it all look easy, but I cannot get it to work (atleast not while Autowiring -- there are probably other uses for such custom editors).

In short, my request is whether or not there is a solution to inject empty collections (instead of null), and possibly to consider adding such a feature in an easily accessible way.

---

**Affects:** 3.2.2

**Comment by spring-issuemaster** [comment]
Bulk closing outdated, unresolved issues. Please, reopen if still relevant.
